src/nbl/video/utilities/CComputeBlit.cpp

```diff
@@ -0,0 +1,152 @@
+#include "nbl/video/utilities/CComputeBlit.h"
+
+using namespace nbl;
+using namespace video;
+
+core::smart_refctd_ptr<video::IGPUSpecializedShader> CComputeBlit::createAlphaTestSpecializedShader(const asset::IImage::E_TYPE inImageType, const core::vectorSIMDu32& workgroupDims,
+	const uint32_t alphaBinCount)
+{
+	char shaderStart[] = "#version 460 core\n#include \"nbl/builtin/glsl/blit/default_compute_alpha_test.comp\"\n";
+	auto shader = core::make_smart_refctd_ptr<asset::CCustomAllocatorCPUBuffer<core::null_allocator<uint8_t>>>(strlen(shaderStart), shaderStart, core::adopt_memory);
+
+	auto cpuShader = core::make_smart_refctd_ptr<asset::ICPUShader>(std::move(shader), asset::ICPUShader::buffer_contains_glsl, asset::IShader::ESS_COMPUTE, "CComputeBlit::createAlphaTestSpecializedShader");
+
+	auto cpuShaderOverriden = asset::IGLSLCompiler::createOverridenCopy(cpuShader.get(),
+		"#define _NBL_GLSL_WORKGROUP_SIZE_X_ %d\n"
+		"#define _NBL_GLSL_WORKGROUP_SIZE_Y_ %d\n"
+		"#define _NBL_GLSL_WORKGROUP_SIZE_Z_ %d\n"
+		"#define _NBL_GLSL_BLIT_DIM_COUNT_ %d\n"
+		"#define _NBL_GLSL_BLIT_ALPHA_BIN_COUNT_ %d\n",
+		inImageType >= asset::IImage::ET_1D ? workgroupDims.x : 1u,
+		inImageType >= asset::IImage::ET_2D ? workgroupDims.y : 1u,
+		inImageType >= asset::IImage::ET_3D ? workgroupDims.z : 1u,
+		static_cast<uint32_t>(inImageType) + 1u,
+		alphaBinCount);
+
+	auto gpuUnspecShader = device->createGPUShader(std::move(cpuShaderOverriden));
+
+	return device->createGPUSpecializedShader(gpuUnspecShader.get(), { nullptr, nullptr, "main" });
+}
+
+core::smart_refctd_ptr<video::IGPUSpecializedShader> CComputeBlit::createNormalizationSpecializedShader(const asset::IImage::E_TYPE inImageType, const asset::E_FORMAT outImageFormat,
+	const asset::E_FORMAT outImageViewFormat, const core::vectorSIMDu32& workgroupDims, const uint32_t alphaBinCount)
```

Comment:
I think you only want to know the `outImageViewFormat`, rest wouldn't matter

Reply:
I use `outImageFormat` to determine whether to insert the define `_NBL_GLSL_BLIT_SOFTWARE_ENCODE_FORMAT_` or not, like:
```C++
if (outImageFormat != outImageViewFormat)
    shaderSourceStream << "#define _NBL_GLSL_BLIT_SOFTWARE_ENCODE_FORMAT_ " << outImageFormat << "\n";
```

That said, I think we can always insert the define. Then in the shader we will have the following two overloads:
```GLSL
vec4 nbl_glsl_blit_encode(in vec4 value)
{
    return value; // a dummy passthrough
}

uvec4 nbl_glsl_blit_encode(in vec4 value)
{
    // do encoding depending upon _NBL_GLSL_BLIT_SOFTWARE_ENCODE_FORMAT_
}
```
Then it would automatically select which overload to use based on the type of `_NBL_GLSL_BLIT_OUT_DESCRIPTOR_DEFINED_` either `imageND` or `uimageND`.

This would work if we can ensure that the encoding will _always_ happen into a uint image format --which I think is a reasonable assumption.

Should we do this?

Reply:
> I  use outImageFormat to determine whether to insert the define _NBL_GLSL_BLIT_SOFTWARE_ENCODE_FORMAT_ or not

what I meant is that for the mipmap creation I care about how I "view" the texture, not what the original format was.

If you need to know what view format needs to be created in order to "store" (what uint format override) then you can use your own function to work it out internally, no need to burden the user.

> Then it would automatically select which overload to use based on the type of _NBL_GLSL_BLIT_OUT_DESCRIPTOR_DEFINED_ either imageND or uimageND.

I forgot GLSL has overloads, then in theory you could define
```glsl
void nbl_glsl_blit_store(in image1DArray target, in ivec2 uv, in vec4 value)
{
   imageStore(target,uv,value);
}
void nbl_glsl_blit_store(in image2DArray target, in ivec3 uv, in vec4 value)
{
   imageStore(target,uv,value);
}
void nbl_glsl_blit_store(in image3D target, in ivec3 uv, in vec4 value)
{
   imageStore(target,uv,value);
}


uvec4 nbl_glsl_blit_software_encode(in vec4 value);

void nbl_glsl_blit_store(in uimage1DArray target, in ivec2 uv, in vec4 value)
{
   imageStore(target,uv,nbl_glsl_blit_software_encode(value));
}
void nbl_glsl_blit_store(in uimage2DArray target, in ivec3 uv, in vec4 value)
{
   imageStore(target,uv,nbl_glsl_blit_software_encode(value));
}
void nbl_glsl_blit_store(in uimage3D target, in ivec3 uv, in vec4 value)
{
   imageStore(target,uv,nbl_glsl_blit_software_encode(value));
}
```

then according to your wishes you can keep all occurences of `_NBL_GLSL_BLIT_SOFTWARE_ENCODE_FORMAT_` handling to the implemenetation of `nbl_glsl_blit_software_encode`

Reply:
I like your overload abuse, its clever.